src/Compiler/Microsoft.AspNetCore.Razor.Language/src/AllowedChildTagDescriptorBuilder.cs

```diff
@@ -1,15 +1,92 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public abstract class AllowedChildTagDescriptorBuilder
+public partial class AllowedChildTagDescriptorBuilder : IBuilder<AllowedChildTagDescriptor>
 {
-    public abstract string Name { get; set; }
+    private static readonly ObjectPool<AllowedChildTagDescriptorBuilder> s_pool = DefaultPool.Create(Policy.Instance);
 
-    public abstract string DisplayName { get; set; }
+    internal static AllowedChildTagDescriptorBuilder GetInstance(TagHelperDescriptorBuilder parent)
+    {
+        var builder = s_pool.Get();
 
-    public abstract RazorDiagnosticCollection Diagnostics { get; }
+        builder._parent = parent;
+
+        return builder;
+    }
+
+    internal static void ReturnInstance(AllowedChildTagDescriptorBuilder builder)
+        => s_pool.Return(builder);
+
+    [AllowNull]
+    private TagHelperDescriptorBuilder _parent;
+    private ImmutableArray<RazorDiagnostic>.Builder? _diagnostics;
+
+    private AllowedChildTagDescriptorBuilder()
+    {
+    }
+
+    internal AllowedChildTagDescriptorBuilder(TagHelperDescriptorBuilder parent)
+    {
+        _parent = parent;
+    }
+
+    public string? Name { get; set; }
+    public string? DisplayName { get; set; }
+
+    public ImmutableArray<RazorDiagnostic>.Builder Diagnostics
+        => _diagnostics ??= ImmutableArray.CreateBuilder<RazorDiagnostic>();
```

Comment:
This, and the others like it, probably needs to be an Interlocked.CompareExchange as well, to make sure multiple threads don't double assign and we lose some diagnostics to the void.

Reply:
Unless, of course, we're confident these builders are never populated in a multi-threaded fashion. I'm unsure whether that's the case today.

Reply:
Note that these thread-safety concerns are not new. They are mostly used locally in compiler passes. I have not seen scenarios where they are shared.

Reply:
Additionally, I don't think it's worth the effort to make builders thread-safe. `ImmutableArray<T>.Builder` isn't thread-safe, so I'm not sure why we hold these to a different standard.

Reply:
Nope, if these aren't intended to be threadsafe, this is fine with me. Just making sure of it.